cpp/src/fil/infer.cu

```diff
@@ -198,8 +201,46 @@ struct finalize_block {
     : tmp_storage(tmp_storage_), num_classes(num_classes_) {}
 
   template <int NITEMS>
-  static size_t smem_footprint() {
-    return sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
+  static size_t smem_footprint(bool predict_proba) {
+    return predict_proba
+             ? sizeof(typename BlockReduceHost<NITEMS>::TempStorage)
+             : sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
+  }
+
+  template <int NITEMS>
+  __device__ __forceinline__ vec<NITEMS, float> shifted_exp(
+    vec<NITEMS, float> margin, vec<NITEMS, float> max) {
+    vec<NITEMS, float> res;
+    for (int i = 0; i < NITEMS; i++) res[i] = expf(margin[i] - max[i]);
+    return res;
+  }
+
+  template <int NITEMS>
+  __device__ __forceinline__ vec<NITEMS, float> block_all_max(
+    vec<NITEMS, float> value, int valid_threads, void* storage = NULL) {
+    if (storage == NULL) storage = tmp_storage;
```

Comment:
It would be better (if it is possible) to either always use the storage in the class, or always pass it explicitly. In the latter case, it's better to move the reduction function outside of `finalize_block`.